src/main/scala/commonist/ui/ImageUI.scala

```diff
@@ -155,20 +155,35 @@ final class ImageUI(file:File, icon:Option[Icon], thumbnailMaxSize:Int, programH
 
 	// BETTER move unparsers and parsers together
 
-	private val	exif		= EXIF extract file
-	private val exifDate	= exif.date		cata ("", _ format "yyyy-MM-dd HH:mm:ss")
-	private val exifGPS		= exif.gps		cata ("", it => it.latitude.toString + "," + it.longitude.toString)
-	private val exifHeading	= exif.heading	cata ("", _.toString)
-	private val exifDesc	= exif.description getOrElse ""
-	private val fixedName	= Filename fix file.getName
-
-	uploadEditor		setSelected	false
-	nameEditor			setText		fixedName
-	descriptionEditor	setText		exifDesc
-	dateEditor			setText		exifDate
-	coordinatesEditor	setText		exifGPS
-	headingEditor		setText		exifHeading
-	categoriesEditor	setText		""
+	private def getExif():ImageData = {
+	  val	exif = EXIF extract file
+	  return new ImageData(
+	      file,
+				false,
+				Filename fix file.getName,
+				exif.description getOrElse "",
+				exif.date		  cata ("", _ format "yyyy-MM-dd HH:mm:ss"),
+				exif.gps		  cata ("", it => it.latitude.toString + "," + it.longitude.toString),
+				exif.heading	cata ("", _.toString),
+				""
+	    )
+	}
+
+	def loadImageData(data:ImageData) {
+  	uploadEditor		  setSelected	data.upload
+  	nameEditor			  setText		data.name
+  	descriptionEditor	setText		data.description
+  	dateEditor			  setText		data.date
+  	coordinatesEditor	setText		data.coordinates
+  	headingEditor		  setText		data.heading
+  	categoriesEditor	setText		data.categories
+	}
+
+	private val exifData:ImageData = getExif
+	// initialize from EXIF data, if any
+	loadImageData(exifData)
+	// initialize from OAI-PMH data, if any
+	oaipmh.map(_.getImageData(exifData)).filter(_.isDefined).map(_.map(loadImageData))
```

Comment:
this looks wrong: `map` shouldn't be used for side-effects, that's what `foreach` is for. instead filtering  for isDefined you can flatten the Options with a `flatMap`. i'd prefer something like
`oaipmh flatMap (_ getImageData exifData) foreach loadImageData`
(the lack of dots and parens being my personal preference)

Reply:
I replaced it by a `Vector` and I get "value flatmap is not a member of Vector" for `oaipmh flatmap (_.getImageData exifData) foreach loadImageData _`

Reply:
I pushed a commit that solves most of the issues. The last ones I fail to answer them, especially this one.